### PR TITLE
[Enhancement] Optimize HeapMergeIterator read speed

### PR DESCRIPTION
**Pull request by xiaoyong-z**
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

#5826

## Problem Summary(Required) ：

The merge_iterator uses the heap to order the chunk, each time it will pop a min_chunk from the heap.
(1) If the min_chunk is less than other chunks in the heap, then it will swap the whole min_chunk out.
(2) If not, it will read only one line from the min_chunk and append it to the resulting chunk, then push this min_chunk back to the heap.

The above (2) procedure is the bottleneck of the merge_iterator. Each time we process one line, we need to append this line, pop, and then push the min chunk to the heap.

To solve this problem, each time, I will get the last row in the min chunk whose key value is smaller than other chunks in the heap, and then append all rows together to the resulting chunk.

In my test, there are 160 million random data. The horizontal base compaction before optimization takes 155 seconds, after optimization it only takes 107 seconds, improving by 1.4486x.

### CODE REVIEW

**Comment by xiaoyong-z** [comment]
# Test case

Table:
```sql
CREATE TABLE `table1` (
  `key1` int NOT NULL COMMENT "",
  `key2` datetime NOT NULL COMMENT "",
  `key3` tinyint NOT NULL COMMENT "",
  `key4` smallint NOT NULL COMMENT "",
  `key5` tinyint NOT NULL COMMENT "",
  `key6` tinyint NOT NULL COMMENT "",
  `key7` tinyint NOT NULL COMMENT "",
  `key8` tinyint NOT NULL COMMENT "",
  `key9` smallint NULL COMMENT "",
  `key10` varchar(20) NOT NULL DEFAULT "baishan" COMMENT "",
  `key11` datetime NOT NULL COMMENT "",
  `key12` datetime NOT NULL COMMENT "",
  `key13` datetime NOT NULL COMMENT "",
  `value1` bigint SUM NULL COMMENT "",
  `value2` bigint SUM NULL COMMENT "",
  `value3` bigint SUM NULL COMMENT "",
  `value4` bigint SUM NULL DEFAULT "0" COMMENT "",
  `update_time` datetime REPLACE NULL COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(`key1`, `key2`, `key3`, `key4`, `key5`, `key6`, `key7`, `key8`, `key9`, `key10`, `key11`, `key12`, `key13`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`key1`) BUCKETS 1
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
);
```
Step1: use python to generate 160 million random data.
Step2: enable cumulative and base compaction, let 160 million data merged into base rowsets.
Step3: Disable cumulative and base compaction, load 60 files, each with around 100000 lines of random data.
Step4: Enable base compaction, monitor the performance of base compaction, which merges 61 rowsets (1 base rowset and 60 incoming rowsets) into a larger base rowset.

**Comment by xiaoyong-z** [comment]
# Results
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

Compaction type | Before optimization, duration of base compaction | After optimization, duration of base compaction
-- | -- | --
Horizontal compaction | 155.06s | 107s

File: be/src/storage/merge_iterator.cpp [hunk 1]
```diff
@@ -64,7 +64,23 @@ class ComparableChunk : public MergingChunk {
...
+        return less_than(last_row, rhs);
+    }
+
+    // return the last row whose key value is less than all values in |rhs|
+    size_t last_row_less_than(const ComparableChunk& rhs, size_t limit_num) {
+        // As we previously pop this chunk from the heap top, `_compared_row` in this chunk
+        // must be less than all rows in rhs, thus here we start comparision from _compared_row + 1;
+        size_t last_row = _compared_row + 1;
+        size_t upper_bound = std::min(_compared_row + limit_num, _chunk->num_rows() - 1);
+        while (last_row <= upper_bound && less_than(last_row, rhs)) {
```
Comment (trueeyu) [inline be/src/storage/merge_iterator.cpp:76]: why not use binary search?

Comment (xiaoyong-z) [reply be/src/storage/merge_iterator.cpp:76]: If all elements are random, min_chunk contains only a few rows (may be less than 10) whose key values are less than all other chunks. In this case, binary search is slower than linear search.

**Review by trueeyu — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Comment by decster** [comment]
run starrocks_be_unittest

**Pull request merged**